lib/mod.lua

```diff
@@ -70,6 +88,10 @@ meta_fake_midi.__index = function(t, key)
                 idx = 1
             end
             if idx <= 16 then
+                if t.real_midi.vports[idx].name == "nb" then
```

Comment:
How does the user use this — do they rename a midi device/port to "nb" from the midi menu system? Can you do that for an unused port?

I have never actually renamed any midi devices.

Reply:
You provided this name on line 7.  (Also on line 79 of this PR, but I don't think that is the operative source of the string.)

When `my_midi` is inserted into the "fake" vports list an entry named "nb" shows up in the list available to the system->devices->midi menu .  You can go to any of the numbered slots in that menu and make it point to "nb" just like you can make the slot point to any of the real midi devices.  Before this PR, setting a slot to "nb" would have no effect, which frankly was confusing to the average user (i.e., me).  With this PR, if a script tries to use that slot for midi, it will be connected to nbout.

This is (I believe) the normal and least surprising way for a user to choose a midi device.  Many scripts do have an internal midi device menu and simply default to using device 1.  You can control what device 1 is by setting slot 1 on the system midi device menu to point where you wish.  With this PR, nbout can participate in this UI.

There are a couple of rough edges, though:

1. Just matching the string "nb" is clumsy.  I don't know enough low-level lua to know if there is a way to safely embed a unique identifier in `my_midi` so that it can be authoritatively matched in the `connect` call, but that would be a more appropriate solution.

2. "nb" doesn't get injected into the vports list until a script is run.  So if you turn on norns and no script is running (like after a restart) and you go straight to the system midi device menu to select "nb" for some script that always uses device 1, there will be no "nb" available.  You have to select the script (which will fail to make any sound because the wrong device is in slot 1), then go choose "nb" on the system menu (because it now exists) and then re-select the script (so that it starts up with a usable midi device in slot 1).  Annoying and confusing.  This rigamarole is mitigated by the fact that norns does not clean up the slots in the system midi device list, so once you put "nb" in slot 1 (or wherever) it will stay there across reboots.

One way to mitigate issue 2 would be to initialize nbout at system startup instead of at script init.  Personally, I would prefer this approach (as well as putting the nbout parameters in the mod menu instead of in the script params) because I typically want to curate the settings on a synth engine separately from working with those sounds in any script/sequencer environment.  It's quite frustrating at the moment to make the same settings in doubledecker over and over as I move from one script to another or try to direct a script with an existing preset to doubledecker instead of to hardware.  I would like nbout to retain its settings across multiple scripts for as long as norns stays on.  Ideally I would also like to save nbout presets, but I don't know to achieve that.  (Hmmm.  Could there be a nbout preset manager script that used its own preset storage to hold values for nbout?  Then you would select that script, load a preset, select the script you want to use nbout with and the nbout settings would remain unchanged!)

Reply:
Oh wiild i think I subtly misunderstood the entire way one handles midi in Norns a bit. Let me know when you’re satisfied with the state of this PR and I’ll take another quick pass and merge it. Thank you for your work.

Reply:
I'm fine with the state of the PR as it is.  Certainly there is more research to be done into the variety of ways that scripts use MIDI, but this gets us over the hump of nbout being broadly available in most scripts that have MIDI out, so I'm happy with that.